src/Orleans/Runtime/CallbackData.cs

```diff
@@ -13,30 +15,51 @@ internal interface ITimebound
         /// This method is called by the timer when the time out is reached.
         /// </summary>
         void OnTimeout();
-        TimeSpan RequestedTimeout();
+
+        DateTime DueTime { get; }
```

Comment:
I would avoid this, since it could mean that each item could change the due time dynamically, or even have different immutable values, which would not be supported by the timer wheel. Instead you could have this as an implementation detail of the timer wheel (basically instead of having a Queue<ITimebound> have a Queue<Tuple<ITimeboud>> (or even better a struct instead of Tuple, and potentially a weak reference to the ITimebound.
With this change, then you can optimize for the idle case too and avoid having a timer in the first place, and have a never ending async task that does `await Task.Delay(time to next expiration in the queue)` (because DueTime is not dynamic and the queue now can enforce an ever increasing due time) or explicitly delay a few milliseconds to yield the thread a little.